### PR TITLE
Canonicalize mux trees

### LINKED ISSUE #900: Lack of mux folding causes unreachable code which tools complain about
```
circuit MuxTree :
  module MuxTree :
    input clock : Clock
    input reset : UInt<1>
    input c1 : UInt<1>
    input c2 : UInt<1>
    input d1 : UInt<5>
    input d2 : UInt<5>
    input d3 : UInt<5>
    output foo : UInt<5>
    

    node baz = mux(c1,d2, d3) 
    node bar = mux(c1, d1, baz) 
    foo <= bar 
```
Gives
```
module MuxTree(
  input        clock, reset, c1, c2,
  input  [4:0] d1, d2, d3,
  output [4:0] foo);

  assign foo = c1 ? d1 : c1 ? d2 : d3; 
endmodule
```
assigning d2 to foo is unreachable and generates warnings.

### PR DESCRIPTION
**Pull request by darthscsi**
closes #900

Will add other cases

### CODE REVIEW

File: include/circt/Dialect/FIRRTL/Canonicalization.td [hunk 1]
```diff
@@ -43,4 +43,16 @@ def GTWithConstLHS : Pat<
...
 
+// mux(cond, x, mux(cond, y, z))
+def MuxSameCondLow : Pat<
+  (MuxPrimOp $cond, $x, (MuxPrimOp $cond, $y, $z)),
+  (MuxPrimOp $cond, $x, $z)
+  >;
+
+// mux(cond, x, node(mux(cond, y, z)))
+def MuxSameCondLowNode : Pat<
+  (MuxPrimOp $cond, $x, (NodeOp (MuxPrimOp $cond, $y, $z), $name, $attr)),
```
Comment (lattner) [inline include/circt/Dialect/FIRRTL/Canonicalization.td:54]: Please don't include Node into this pattern.  We should separately eliminate any node operations (or drop them from the firrtl dialect entirely).

**Review by lattner — APPROVED** [review]
Looks good without the `node` part of the pattern.  We should eliminate nodes that have no annotations on them as their own canonicalization.

File: include/circt/Dialect/FIRRTL/Canonicalization.td [hunk 2]
```diff
@@ -43,4 +43,17 @@ def GTWithConstLHS : Pat<
   (LTPrimOp $rhs, $lhs),
   [(NonConstantOp $rhs)]>;
 
+// mux(cond, x, mux(cond, y, z))
```
Comment (seldridge) [inline include/circt/Dialect/FIRRTL/Canonicalization.td:46]: Nit: 
```suggestion
// mux(cond, x, mux(cond, y, z)) -> mux(cond, x, z)
```

File: include/circt/Dialect/FIRRTL/Canonicalization.td [hunk 3]
```diff
@@ -43,4 +43,17 @@ def GTWithConstLHS : Pat<
...
+def MuxSameCondLow : Pat<
+  (MuxPrimOp $cond, $x, (MuxPrimOp $cond, $y, $z)),
+  (MuxPrimOp $cond, $x, $z)
+  >;
+
+// mux(cond, mux(cond, y, z), x)
```
Comment (seldridge) [inline include/circt/Dialect/FIRRTL/Canonicalization.td:52]: Nit: 
```suggestion
// mux(cond, mux(cond, y, z), x) -> mux(cond, y, x)
```

**Review by seldridge — APPROVED** [review]
lgtm

I'm good to see where this solution takes us. It would be good to get user feedback at some point.

**Pull request merged**